server/src/test/java/org/opensearch/index/engine/InternalEngineTests.java

```diff
@@ -7546,6 +7546,89 @@ public void testGetSegmentInfosSnapshot() throws IOException {
         engine.close();
     }
 
+    // This test performs indexing ops and verifies engine.getSegmentInfosSnapshot() returns SegmentInfos having highest
+    // generation
+    public void testGetSegmentInfosSnapshotWithIndexing() throws IOException {
+        final int numDocs = randomIntBetween(10, 100);
+        for (int docId = 0; docId < numDocs; docId++) {
+            index(engine, docId);
+            if (randomBoolean()) {
+                engine.refresh("test");
+            }
+        }
+        engine.flush(true, true);
+
+        // Fetch SegmentInfos from memory & disk
+        final SegmentInfos inMemorySegInfos = engine.getLatestSegmentInfos();
+        final long generation = inMemorySegInfos.getGeneration();
+        final long lastCommitGeneration = SegmentInfos.getLastCommitGeneration(store.directory());
+
+        // Verify largest (generation file) of two is returned in engine.getSegmentInfosSnapshot
+        SegmentInfos segmentInfos = engine.getSegmentInfosSnapshot().get();
```

Comment:
I think this would need to be surrounded by try-with-resources so the snapshot is closed or the directory won't close?